### ISSUE TITLE
since v0.91 last character of nvarchar chopped when string is at max length and contains unicode chars

### ISSUE THREAD

**Issue by beaudet** [FreeTDS/freetds#94]
To reproduce: create table with nvarchar(128) and insert the following string which contains unicode character "ì"

Soprintendenza per il patrimonio storico artistico e demoetnoantropologico per Bologna, Ferrara,, Forlì-Cesena, Ravenna e Rimini

In patched 0.82, this string is properly returned and length reported as 128.  In version 0.91 and newer, length is reported as 129, the last "i" in the string (not the unicode char) is chopped off and the last byte is the null character.

RHEL6, 32 bit, Perl 5 - freetds and freetds-unixodbc packages built using rpmbuild.  Can reproduce problem using isql utility in addition to the following code:

#!/usr/bin/perl
use warnings;
use strict;
use DBI;
use Error qw(:try);
use MIME::Lite;
use Data::Dumper;
use Sys::Hostname;
use Time::HiRes qw(usleep);
my $dbh;

my $cnt=0;
while (++$cnt) {  // loop used for testing another problem with intermittent TLS errors

    try {
        print "$cnt:\ttrying connection\n";
        $dbh = DBI->connect('dbi:ODBC:MYDB',myuser', 'mypassword');
        die 'no dbh' if ! $dbh;
        $dbh->{LongTruncOk}=1;
        my $inst = $dbh->selectrow_array('SELECT mystring FROM mytable WHERE id=12345');
        my $char = chop $inst;
        die "bug with unicode exists" if $char ne "i";
    }
    catch Error with {
        my $details = "";
        $details = "Can't connect to database: $DBI::errstr!" if defined $DBI::errstr;
        my $exception = shift;
        $details = $details."\n".Dumper($exception);
        print "DB error: $details\n";
    }
    finally {
        $dbh->disconnect if $dbh;;
        # sleep for a short time
        usleep(10000);
    };
}

**Comment by freddy77** [comment]
Can you provide an ODBC trace?

**Comment by beaudet** [comment]
Yes, will send tomorrow.

-------- Original message --------

**Comment by freddy77** [comment]
This actually could be expected if the client encoding is utf-8. In this case the string would require 129 bytes (so the output length would be 129 and not 128) but if DBD::ODBC is passing a buffer of 129 bytes (128 + 1 for the termination) only 128 bytes (one is consumed by the terminator) are returned, that is the entire string converted without the last character (as there is no more space).

**Comment by beaudet** [comment]
My understanding is that the SQL maximum lengths defined for nvarchar in MSSQL are maximum string lengths as measured by character count rather than byte count.  A variable byte length encoding like UTF-8 could therefore have a byte length that does not match the column definition.  And that would be different from the internal SQL Server data length which seems to be fixed at two bytes per character although I don't think that would matter much to DB clients.

TSQL and ISQL give different results.  TSQL always seems to return the full 128 bytes while ISQL has this behavior when freetds 0.82 is installed, but not with versions of freetds > 0.82.

I agree the output length would be 129 and not 128, but why would DBD:ODBC be passing a buffer that is too short?  Shouldn't it be 130 = utf-8 data length of 129 + 1 for termination character in this case?  And why would DBD:ODBC work with freetds 0.82 but not freetds > 0.82?

**Comment by beaudet** [comment]
I hope to provide an update on this next week.

**Comment by freddy77** [comment]
From an ODBC trace you could check which length DBD::ODBC is passing to FreeTDS.
Also you should see which information DBD::ODBD retrieve to compute that length.
Is DBD::ODBC using wide version (if I remember must be compiled with a different option) ?

**Comment by beaudet** [comment]
I just tried to reproduce this issue on a 32 bit RHEL6 system with the latest 1.00.24 patch version - not sure the prior 1.00.24 version I was using was the patched version.  Anyway, this test was on a system that wasn't previously using DBD:ODBC or any ODBC connections for that matter, and now it doesn't seem to be happening.  I installed the RHEL provided DBD:ODBC and I'm wondering if that has anything to do with it.  I'm going to try rebuild using 1.00.24 patched on the system that was originally experiencing the problem and will report back.
